### DISCUSSION TITLE
What are the Recommended Ruff configurations to avoid conflicts with SQLA-2.x typing

### DISCUSSION THREAD

**Discussion by ashish16052** [sqlalchemy/sqlalchemy#10980] [Usage Questions]
I've encountered disagreements between Ruff linting and the new SQLAlchemy declarative typing. Here are some specific instances:

1. Handling nullable variable length strings:
    SQLAlchemy (SQLA) suggests: name: Mapped[Optional[str]] (as per documentation)
    Ruff recommends: name: Mapped[str | None] (which SQLA doesn't support)

2. Handling relationships:
   SQLA suggests: Mapped["Parent"] (assuming Parent model is in another module not yet imported)
   Ruff conflicts with rules UP037 and F821.

I'm seeking recommendations on how to address these conflicts and ensure compatibility between Ruff linting and SQLAlchemy 2.x typing. Additionally, I'd appreciate guidance on best practices for handling similar conflicts in other scenarios.

**Comment by CaselIT** [comment]
Hi

> Ruff recommends: name: Mapped[str | None] (which SQLA doesn't support)

that's fully supported by sqlalchemy. It's not documented using `str | None` since py3.8 and py3.7 do not support it.

> SQLA suggests: Mapped["Parent"] (assuming Parent model is in another module not yet imported)

You would need to import it in an `if TYPE_CHECKING` block.
That's a classic pattern when using python typing to resolve circular references, it's not something that's only in sqlalchemy

**Comment by zzzeek** [comment]
> I've encountered disagreements between Ruff linting and the new SQLAlchemy declarative typing. Here are some specific instances:
> 
>     1. Handling nullable variable length strings:
>        SQLAlchemy (SQLA) suggests: name: Mapped[Optional[str]] (as per documentation)
>        Ruff recommends: name: Mapped[str | None] (which SQLA doesn't support)

SQLAlchemy fully supports `Mapped[str | None]`

 
>     2. Handling relationships:
>        SQLA suggests: Mapped["Parent"] (assuming Parent model is in another module not yet imported)
>        Ruff conflicts with rules UP037 and F821.

Use `Mapped[Parent]` without the quotes.   Use [pep-563](https://peps.python.org/pep-0563/#enabling-the-future-behavior-in-python-3-7):   `from __future__ import annotations`


> 
> 
> I'm seeking recommendations on how to address these conflicts and ensure compatibility between Ruff linting and SQLAlchemy 2.x typing. Additionally, I'd appreciate guidance on best practices for handling similar conflicts in other scenarios.
